### DISCUSSION TITLE
How can I reload KML file and its updated data every 5 seconds - so it will be reflected on the TerriaMaps view.

### DISCUSSION THREAD

**Discussion by eliadd** [TerriaJS/terriajs#7231] [Q&A] [answered]
Hi,

I am experimenting with the TerriaJS & TerriaMaps solutions.
In my POC I have a KML file with cars that are dynamically changed and therefore the number of cars need to be changed on the TerriaMaps view.

Still, I can only see the initial state of the first KML data and later changes on the KML file are not reflected in the TerriaMaps view.

So my Q - how can I reload the KML file and its new data every 5 seconds - so it will be reflected on the TerriaMaps view.

If you have any other approach to solve this issue - please feel free to elaborate.

Thanks,

**Comment by na9da** [comment]
Hi @eliadd, 

Terria has an [auto-refresh feature](https://github.com/TerriaJS/terriajs/blob/fae976658debb09c73cc1a9132a6f70a48c25f06/lib/Traits/TraitsClasses/AutoRefreshingTraits.ts#L11) that works for some catalog items, eg: [CSV](http://ci.terria.io/main/#share=s-vLRXfJtqzffRc6id5hgRq9v29NI), CZML - this is currently not implemented for KML.

A simple implementation would be to reload the whole item at specific intervals, but this would be expensive and causes a flicker like in the screencast below. 

![kml-refresh](https://github.com/TerriaJS/terriajs/assets/1430/a85dd3bf-5eeb-4eb9-8921-c0e7ae0ef822)


The underlying Cesium datasource supports an [update](https://cesium.com/learn/cesiumjs/ref-doc/KmlDataSource.html?classFilter=kmlda#update) method which allows updating the network links in a KML file. When used with an appropriate [refreshInterval](https://developers.google.com/kml/documentation/kmlreference#example_6) it might be possible to refresh more efficiently (without reloading the whole dataset). 

Does your KML use network links? If so are you able to share a sample for testing?

**Reply by eliadd** [reply]
Thx a lot for the quick response Nanda!
Currently in my POC I have local and static KML file that I am generating & put it in local directory for uploads to TerriaMaps (with no network-links).

Can you please elaborate if & where I need to update TerriaJS/TerriaMaps to reload the KML in both cases you mention:
1. Reload the whole item at specific intervals - can you please share the code you demonstrate on the screencast above?
2. Updating the underlying Cesium datasource that supports an update method which allows updating dynamically the KML resource in TerriaMaps.

I will first try the first option in my POC - if it will be poor experiance from performance POV - then, I will move forword to the second option.

Thx again!

Eliad.

**Reply by na9da** [reply]
Here's a link to [see it in action](http://ci.terria.io/test-autorefresh-kml/#share=s-AkrRQ8Eok28oByrqJgl7z6J840N&clean&https://gist.githubusercontent.com/na9da/e6a592f4d456c545fccbc1d5c4910502/raw/e9b71607582b3ebac5af0d867ab9770d9295efaa/kml-refresh.json).

This required changes to terriajs - I have pushed these to a separate branch of terriajs - [test-autorefresh-kml](https://github.com/terriaJS/terriajs/tree/test-autorefresh-kml) - you can  check out the [diff](https://github.com/TerriaJS/terriajs/compare/test-autorefresh-kml) to see what these changes are. The catalog configuration for that example looks like [this](https://gist.github.com/na9da/e6a592f4d456c545fccbc1d5c4910502).

To test these changes with your map you'll need to [build and run terriamap](https://docs.terria.io/guide/contributing/development-environment/) against the modified terriajs branch.

**Reply by na9da** [reply]
Also, if you have the choice of using a different format, you could try generating GeoJSON or CZML both of which can be refreshed in Terria.

**Reply by eliadd** [reply]
Great Thx Nanada!

I updated my code according your answer and it worked fine - with static KML file - Thx again!

In addition & in order to improve performance - can you please elaborate how can I integrate terriajs/terriaMap with stream of data (with KML format) - as you mention in your previus answer - update internal Cesium with network links (of KML stream) - where do I need to update the code in terriajs/terriaMap (will be great if you could update with code example as you did in your prevoius answer).

Thx again!

Eliad.

**Reply by na9da** [reply]
I did some more testing and looks like the changes to TerriaJS I suggested previously are unnecessary - sorry about that. Cesium will autorefresh the KMLs with correct refresh settings without us having to do anything. Here is an example.

You can write a `main.kml` with the following content and link to `someother.kml` with `refreshInterval` set to the number of seconds.
```xml
<?xml version="1.0" encoding="UTF-8"?>
<NetworkLink id="link">
  <Link>
    <href>./someother.kml</href>
    <refreshMode>onInterval</refreshMode>
    <refreshInterval>10</refreshInterval>
  </Link>
</NetworkLink>
``` 

You can then remove the ` "refreshInterval": 5` from Terria's [json configuration](https://gist.github.com/na9da/e6a592f4d456c545fccbc1d5c4910502#file-kml-refresh-json-L7).

The refresh happens periodically, but there won't be any Terria UI indication of this happening, you need to check your network tab to verify that the `someother.kml` file is getting re-fetched.

Here's more [documentation](https://developers.google.com/kml/documentation/kmlreference#refreshmode) on the KML attributes - however Cesium's support might be limited.

**Reply by eliadd** [reply]
Thx again Nanda,

I did the NT link and it works fine - still, when refreash it is flickering a bit....

I also tried the following configuration on the KML NT link :

<refreshMode>onChange</refreshMode>
<viewRefreshMode>onStop</viewRefreshMode>
<viewRefreshTime>7</viewRefreshTime>

But it didnt work for me - this we be my best option for the POC - so TerriaMap view will be changed when remote model changed - will be great if you could check it out :-).

Also, regarding the great feature of TerriaMap story (crating snapshots from TerriaMap view) - how can I save the story I created? - once refreashed - the story is being deleted.

Thx again for your great support!!!

Eliad.

**Reply by eliadd** [reply]
`<refreshMode>onChange</refreshMode>`
`<viewRefreshMode>onStop</viewRefreshMode>`
`<viewRefreshTime>7</viewRefreshTime>`

**Reply by eliadd** [reply]
-> I also tried the following configuration on the KML NT link :

**Reply by na9da** [reply]
@eliadd, I think some of your messages might have got clipped.

From the Cesium code for [KmlDataSource](https://github.com/CesiumGS/cesium/blob/7b93161da1cc03bdc796b204e7aa51fb7acebf04/packages/engine/Source/DataSources/KmlDataSource.js#L881-L886), it looks like some of these properties are not supported. You should see a warning in the console for the unsupported ones.

> Also, regarding the great feature of TerriaMap story (crating snapshots from TerriaMap view) - how can I save the story I created? - once refreashed - the story is being deleted.

After you have created the story, you can create as share link for it, which will serialize all the inputs - open the share link in another tab to see if it works. 

![image](https://github.com/TerriaJS/terriajs/assets/1430/598a26a6-986a-49d4-8f66-90fa869c96cc)

Note that if you added local files to your story, they wont be part of the share link.

You can also optionally enable URL shortening for share links by configuring an [s3 bucket in your serverconfig.json](https://github.com/TerriaJS/terriajs-server/blob/2c48621ffdf36a7a764bfb981f75820781e383b9/serverconfig.json.example#L134-L152) to save the share links.

**Reply by eliadd** [reply, marked as answer]
Thx a lot Nanda!

You have a great product & great support!

Eliad.
